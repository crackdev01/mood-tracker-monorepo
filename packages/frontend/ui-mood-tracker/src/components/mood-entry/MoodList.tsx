import React from 'react';
import { useTranslation } from 'react-i18next';
import { Button, Header, List } from 'semantic-ui-react';

function MoodList() {
  const { t } = useTranslation(['MoodEntry']);

  return (
    <article className="mood-entry__list">
      <Header as="h3">{t('entries')}</Header>

      <List divided verticalAlign="middle">
        <List.Item>
          <List.Content floated="right">
            <Button basic color="grey">
              {t('buttons.edit')}
            </Button>

            <Button basic color="red">
              {t('buttons.delete')}
            </Button>
          </List.Content>

          <List.Content>Entry 1</List.Content>
        </List.Item>

        <List.Item>
          <List.Content floated="right">
            <Button basic color="grey">
              {t('buttons.edit')}
            </Button>

            <Button basic color="red">
              {t('buttons.delete')}
            </Button>
          </List.Content>

          <List.Content>Entry 2</List.Content>
        </List.Item>

        <List.Item>
          <List.Content floated="right">
            <Button basic color="grey">
              {t('buttons.edit')}
            </Button>

            <Button basic color="red">
              {t('buttons.delete')}
            </Button>
          </List.Content>

          <List.Content>Entry 3</List.Content>
        </List.Item>
      </List>
    </article>
  );
}

export default MoodList;
