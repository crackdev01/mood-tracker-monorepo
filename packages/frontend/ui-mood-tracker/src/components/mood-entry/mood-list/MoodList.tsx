import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import dayjs from 'dayjs';
import { Header, Icon, Pagination, Table } from 'semantic-ui-react';

import EditEntryModal from '../../../components/mood-entry/edit-entry/EditEntryModal';
import DeleteEntryModal from '../../../components/mood-entry/delete-entry/DeleteEntryModal';
import { ApplicationState } from '../../../store';
import { Mood } from '../../../mood.types';

const MoodList = () => {
  const { t } = useTranslation(['MoodEntry']);
  const moods = useSelector((state: ApplicationState) => state.moodReducer.mood);
  const DEFAULT_PAGE_ENTRIES = 5;
  const [currentPage, setCurrentPage] = useState(1);
  const [visibleEntries, setVisibleEntries] = useState([]);
  const [showEditEntryModal, setShowEditEntryModal] = useState(false);
  const [showDeleteEntryModal, setShowDeleteEntryModal] = useState(false);
  const [editableMood, setEditableMood] = useState({});
  const [deletableMood, setDeletableMood] = useState({});
  const pages = Math.ceil(moods.length / DEFAULT_PAGE_ENTRIES);

  const updatePage = (_: any, data: any) => {
    console.log('updating page');
    setCurrentPage(Math.ceil(data.activePage));
  };

  const updateVisibleEntries = () => {
    let min;
    let max;
    if (currentPage === 1) {
      min = 0;
      max = DEFAULT_PAGE_ENTRIES;
    } else {
      max = currentPage * DEFAULT_PAGE_ENTRIES;
      min = max - DEFAULT_PAGE_ENTRIES;
    }
    setVisibleEntries(moods.slice(min, max));
  };

  const editEntry = (mood: any) => {
    setEditableMood(mood);
    setShowEditEntryModal(true);
  };

  const closeEditModal = () => setShowEditEntryModal(false);

  const deleteEntry = (mood: Mood) => {
    setDeletableMood(mood);
    setShowDeleteEntryModal(true);
  };

  const closeDeleteModal = () => setShowDeleteEntryModal(false);

  useEffect(() => {
    updateVisibleEntries();
  }, [currentPage, moods]);

  return (
    <article className="mood-entry__list">
      <Header as="h3">{t('entries')}</Header>

      <Table singleLine className="mood-entry__list__table">
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>{t('table.headers.enteredAt')}</Table.HeaderCell>
            <Table.HeaderCell>{t('table.headers.status')}</Table.HeaderCell>
            <Table.HeaderCell>{t('table.headers.intensity')}</Table.HeaderCell>
            <Table.HeaderCell />
            <Table.HeaderCell />
          </Table.Row>
        </Table.Header>

        <Table.Body>
          {visibleEntries.map((mood: Mood) => {
            return (
              <Table.Row key={mood.mood_id}>
                <Table.Cell>{dayjs(mood.mood_enteredAt).format('MMM D, YYYY h:mm A')}</Table.Cell>
                <Table.Cell>{t(`moodStatus.${mood.mood_status}`)}</Table.Cell>
                <Table.Cell>{mood.mood_intensity}</Table.Cell>
                <Table.Cell>
                  <Icon name="pencil" size="large" onClick={() => editEntry(mood)} />
                </Table.Cell>
                <Table.Cell>
                  <Icon name="delete" color="red" size="large" onClick={() => deleteEntry(mood)} />
                </Table.Cell>
              </Table.Row>
            );
          })}
        </Table.Body>

        <Table.Footer>
          <Table.Row>
            <Table.HeaderCell colSpan="5">
              <Pagination
                floated="right"
                defaultActivePage={currentPage}
                totalPages={pages}
                onPageChange={updatePage}
              />
            </Table.HeaderCell>
          </Table.Row>
        </Table.Footer>
      </Table>

      <EditEntryModal
        displayModal={showEditEntryModal}
        closeModal={closeEditModal}
        mood={editableMood}
      />

      <DeleteEntryModal
        displayModal={showDeleteEntryModal}
        closeModal={closeDeleteModal}
        mood={deletableMood}
      />
    </article>
  );
};

export default MoodList;
