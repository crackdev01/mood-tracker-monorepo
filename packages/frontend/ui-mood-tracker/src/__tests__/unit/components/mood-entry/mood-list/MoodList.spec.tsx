import React from 'react';
import { useSelector } from 'react-redux';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import MoodList from '../../../../../components/mood-entry/mood-list/MoodList';
import { mockMoods } from '../../../../fixtures';

Enzyme.configure({ adapter: new Adapter() });

const useSelectorMock = useSelector as jest.Mock<any>;
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

describe('MoodList', () => {
  let w: any;
  let useEffectMock;
  beforeEach(() => {
    useSelectorMock.mockReturnValueOnce(mockMoods);
    // FIXME: useEffect is not being triggered.
    useEffectMock = jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
    w = shallow(<MoodList />);
  });

  test('renders', () => {
    expect(w.debug()).toMatchSnapshot();
  });
});
