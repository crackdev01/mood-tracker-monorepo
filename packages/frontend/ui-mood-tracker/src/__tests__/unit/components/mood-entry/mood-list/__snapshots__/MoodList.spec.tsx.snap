// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoodList renders 1`] = `
"<article className=\\"mood-entry__list\\">
  <Header as=\\"h3\\">
    entries
  </Header>
  <Table singleLine={true} className=\\"mood-entry__list__table\\" as=\\"table\\">
    <TableHeader as=\\"thead\\">
      <TableRow as=\\"tr\\" cellAs=\\"td\\">
        <TableHeaderCell as=\\"th\\">
          table.headers.enteredAt
        </TableHeaderCell>
        <TableHeaderCell as=\\"th\\">
          table.headers.status
        </TableHeaderCell>
        <TableHeaderCell as=\\"th\\">
          table.headers.intensity
        </TableHeaderCell>
        <TableHeaderCell as=\\"th\\" />
        <TableHeaderCell as=\\"th\\" />
      </TableRow>
    </TableHeader>
    <TableBody as=\\"tbody\\" />
    <TableFooter as=\\"tfoot\\">
      <TableRow as=\\"tr\\" cellAs=\\"td\\">
        <TableHeaderCell colSpan=\\"5\\" as=\\"th\\">
          <Pagination floated=\\"right\\" defaultActivePage={1} totalPages={3} onPageChange={[Function: updatePage]} aria-label=\\"Pagination Navigation\\" boundaryRange={1} ellipsisItem=\\"...\\" firstItem={{...}} lastItem={{...}} nextItem={{...}} pageItem={{...}} prevItem={{...}} siblingRange={1} />
        </TableHeaderCell>
      </TableRow>
    </TableFooter>
  </Table>
  <EditEntryModal displayModal={false} closeModal={[Function: closeEditModal]} mood={{...}} />
  <DeleteEntryModal displayModal={false} closeModal={[Function: closeDeleteModal]} mood={{...}} />
</article>"
`;
